s3fs/core.py

```diff
@@ -2331,41 +2331,31 @@ def _upload_chunk(self, final=False):
             and self.tell() < self.blocksize
         ):
             # only happens when closing small file, use on-shot PUT
-            data1 = False
+            pass
         else:
             self.buffer.seek(0)
-            (data0, data1) = (None, self.buffer.read(self.blocksize))
-
-        while data1:
-            (data0, data1) = (data1, self.buffer.read(self.blocksize))
-            data1_size = len(data1)
-
-            if 0 < data1_size < self.blocksize:
-                remainder = data0 + data1
-                remainder_size = self.blocksize + data1_size
-
-                if remainder_size <= self.part_max:
-                    (data0, data1) = (remainder, None)
-                else:
-                    partition = remainder_size // 2
-                    (data0, data1) = (remainder[:partition], remainder[partition:])
 
-            part = len(self.parts) + 1
-            logger.debug("Upload chunk %s, %s" % (self, part))
+            # previously last small chunk was merged into previous one 
+            # (and in some cases two chunks were equalized)
+            # However, R2 imposes additional restriction that all but last chunk
+            # are of same size, and last chunk can't be larger.
+            while data := self.buffer.read(self.blocksize):
```

Comment:
The first use of a walrus in this repo. Remind me, which version of python does this require?

Reply:
3.8 (see https://docs.python.org/3/whatsnew/3.8.html)